src/contracts/pods/EigenPod.sol

```diff
@@ -485,256 +505,126 @@ contract EigenPod is IEigenPod, Initializable, ReentrancyGuardUpgradeable, Eigen
         activeValidatorCount++;
         validatorInfo.status = VALIDATOR_STATUS.ACTIVE;
         validatorInfo.validatorIndex = validatorIndex;
-        validatorInfo.mostRecentBalanceUpdateTimestamp = oracleTimestamp;
+        validatorInfo.mostRecentBalanceUpdateTimestamp = beaconTimestamp;
+        validatorInfo.restakedBalanceGwei = restakedBalanceGwei;
 
-        if (validatorEffectiveBalanceGwei > MAX_RESTAKED_BALANCE_GWEI_PER_VALIDATOR) {
-            validatorInfo.restakedBalanceGwei = MAX_RESTAKED_BALANCE_GWEI_PER_VALIDATOR;
-        } else {
-            validatorInfo.restakedBalanceGwei = validatorEffectiveBalanceGwei;
-        }
         _validatorPubkeyHashToInfo[validatorPubkeyHash] = validatorInfo;
 
         emit ValidatorRestaked(validatorIndex);
-        emit ValidatorBalanceUpdated(validatorIndex, oracleTimestamp, validatorInfo.restakedBalanceGwei);
+        emit ValidatorBalanceUpdated(validatorIndex, beaconTimestamp, restakedBalanceGwei);
 
-        return validatorInfo.restakedBalanceGwei * GWEI_TO_WEI;
+        return restakedBalanceGwei * GWEI_TO_WEI;
     }
 
-    function _verifyBalanceUpdate(
-        uint64 oracleTimestamp,
-        uint40 validatorIndex,
+    function _verifyCheckpointProof(
+        uint64 beaconTimestamp,
         bytes32 beaconStateRoot,
-        bytes calldata validatorFieldsProof,
-        bytes32[] calldata validatorFields
-    ) internal returns(int256 sharesDeltaGwei){
-        uint64 validatorEffectiveBalanceGwei = validatorFields.getEffectiveBalanceGwei();
-        bytes32 validatorPubkeyHash = validatorFields.getPubkeyHash();
-        ValidatorInfo memory validatorInfo = _validatorPubkeyHashToInfo[validatorPubkeyHash];
+        BeaconChainProofs.BalanceProof calldata proof
+    ) internal returns (int256 balanceDeltaGwei) {
+        ValidatorInfo memory validatorInfo = _validatorPubkeyHashToInfo[proof.pubkeyHash];
 
-        // 1. Balance updates should be more recent than the most recent update
         require(
-            validatorInfo.mostRecentBalanceUpdateTimestamp < oracleTimestamp,
-            "EigenPod.verifyBalanceUpdate: Validators balance has already been updated for this timestamp"
+            validatorInfo.status == VALIDATOR_STATUS.ACTIVE,
+            "EigenPod._verifyCheckpointProof: validator must be ACTIVE"
         );
 
-        // 2. Balance updates should only be performed on "ACTIVE" validators
+        // Ensure we aren't proving a validator twice for the same checkpoint. This will fail if:
+        // - validator submitted twice during this checkpoint
+        // - validator withdrawal credentials verified after checkpoint starts, then submitted
+        //   as a checkpoint proof
+        // TODO - we might want to "skip" and emit an event, rather than revert?
```

Comment:
I think we should skip here instead of reverting. emitting an event in the case of skipping seems...reasonable, at least.
my thoughts: if we revert, then there is a potential griefing vector, whereby an attacker can cause a large transaction (proving e.g. 50 balances) to revert, by submitting their own transaction that just proves a single balance.
I don't think it's a particularly realistic griefing concern, but I'd rather just not have the attack vector present at all.
Skipping would in general also resolve race conditions more elegantly, IMO.

Reply:
i hadn't actually considered this griefing vector, that's a great point in favor of skipping!

was more thinking this might accidentally occur, but you make a good point that this could be intentionally abused to make proving a checkpoint awful.

Reply:
Fleshing out the griefing idea - let's say you have 1000 validators to prove for your checkpoint, and block gas limit says you can prove them in batches of 25 per tx.

Someone can frontrun any one of these batches with a single proof to cause your whole batch to fail, and they can continue to do this until your checkpoint is complete.

Reply:
Sounds like a nightmare. Let's skip :D 

@gpsanant any input? should we skip + event? or just skip?

Reply:
skip is a good idea.

event is also good, but not needed for backend.